### PR TITLE
toon: avoid fixed dependency from openblas

### PR DESCRIPTION
**Pull request by elvisdukaj**
Specify library name and version:  **toon/***

TooN is a header only library. TooN can be used with or without lapack. The conan recipe was forcing to use always lapack even if it's not mandatory.

According to the official [documentation](https://github.com/edrosten/TooN):

> If you use any LAPACK based features (SVD.h, LU.h, QR_Lapack.h, SymEigen.h, Lapack_Cholesky.h) you will need to link against LAPACK, probably using -llapack and perhaps -lblas. 

I made this dependency optional but enabled by default.

I tested using "Android" where `OpenBLAS` doesn't compile on `conan-center-index` with `flack` disabled. And on Windows with both this option enabled and disabled.

Logs when `with_lapack=False`:

```
Profile host:
[settings]
arch=x86_64
build_type=Release
compiler=msvc
compiler.cppstd=17
compiler.runtime=dynamic
compiler.runtime_type=Release
compiler.version=193
os=Windows
[options]
with_lapack=False
[tool_requires]
*: ninja/1.11.1, cmake/3.28.1
[conf]
tools.cmake.cmaketoolchain:generator=Ninja Multi-Config
tools.env.virtualenv:powershell=False

======== Computing dependency graph ========
Graph root
    cli
Requirements
    toon/3.2#2ec368e97ee6412fe866134e728f8a83 - Cache
Build requirements
    cmake/3.28.1#1a5bd0a27293689fd6fa44584e8e2322 - Cache
    ninja/1.11.1#cd30ca226f2040a979c359ebca4817d5 - Cache

[...]
======== Testing the package: Building ========
toon/3.2 (test package): Calling build()
toon/3.2 (test package): Running CMake.configure()
toon/3.2 (test package): RUN: cmake -G "Ninja Multi-Config" -DCMAKE_TOOLCHAIN_FILE="generators/conan_toolchain.cmake" -DCMAKE_INSTALL_PREFIX="C:/qxr/repo/conan-center-index/recipes/toon/all/test_package" -DCMAKE_POLICY_DEFAULT_CMP0091="NEW" "C:/qxr/repo/conan-center-index/recipes/toon/all/test_package"
conanvcvars.bat: Activating environment Visual Studio 17 - amd64 - winsdk_version=None - vcvars_ver=14.3
[vcvarsall.bat] Environment initialized for: 'x64'
-- Using Conan toolchain: C:/qxr/repo/conan-center-index/recipes/toon/all/test_package/build/msvc-193-x86_64-17-release/generators/conan_toolchain.cmake
-- Conan toolchain: C++ Standard 17 with extensions OFF
-- The CXX compiler identification is MSVC 19.39.33523.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/2022/Professional/VC/Tools/MSVC/14.39.33519/bin/Hostx64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found PkgConfig: C:/Users/edukaj/.conan2/p/b/pkgco8e02b502df764/p/bin/pkgconf.exe (found version "2.1.0") 
-- Checking for module 'TooN'
--   Found TooN, version 3.2
-- Configuring done (3.0s)
-- Generating done (0.0s)
-- Build files have been written to: C:/qxr/repo/conan-center-index/recipes/toon/all/test_package/build/msvc-193-x86_64-17-release

toon/3.2 (test package): Running CMake.build()
toon/3.2 (test package): RUN: cmake --build "C:\qxr\repo\conan-center-index\recipes\toon\all\test_package\build\msvc-193-x86_64-17-release" --config Release -- -j8
conanvcvars.bat: Activating environment Visual Studio 17 - amd64 - winsdk_version=None - vcvars_ver=14.3
[vcvarsall.bat] Environment initialized for: 'x64'
[2/2] Linking CXX executable Release\test_package.exe


======== Testing the package: Executing test ========
toon/3.2 (test package): Running test()
toon/3.2 (test package): RUN: Release\test_package
determinant: 4.61
```

Logs when `with_lapack=True`:

```

======== Input profiles ========
Profile host:
[settings]
arch=x86_64
build_type=Release
compiler=msvc
compiler.cppstd=17
compiler.runtime=dynamic
compiler.runtime_type=Release
compiler.version=193
os=Windows
[options]
with_lapack=True
[tool_requires]
*: ninja/1.11.1, cmake/3.28.1
[conf]
tools.cmake.cmaketoolchain:generator=Ninja Multi-Config
tools.env.virtualenv:powershell=False

======== Computing dependency graph ========
Graph root
    cli
Requirements
    openblas/0.3.26#5e63df7a303d8f591b34f1fb4d7fb2d2 - Cache
    toon/3.2#2ec368e97ee6412fe866134e728f8a83 - Cache
Build requirements
    cmake/3.28.1#1a5bd0a27293689fd6fa44584e8e2322 - Cache
    ninja/1.11.1#cd30ca226f2040a979c359ebca4817d5 - Cache

[...]

======== Testing the package ========
Removing previously existing 'test_package' build folder: C:\qxr\repo\conan-center-index\recipes\toon\all\test_package\build\msvc-193-x86_64-17-release
toon/3.2 (test package): Test package build: build\msvc-193-x86_64-17-release
toon/3.2 (test package): Test package build folder: C:\qxr\repo\conan-center-index\recipes\toon\all\test_package\build\msvc-193-x86_64-17-release
toon/3.2 (test package): Writing generators to C:\qxr\repo\conan-center-index\recipes\toon\all\test_package\build\msvc-193-x86_64-17-release\generators
toon/3.2 (test package): Generator 'PkgConfigDeps' calling 'generate()'
toon/3.2 (test package): Generator 'VirtualRunEnv' calling 'generate()'
toon/3.2 (test package): Calling generate()
toon/3.2 (test package): Generators folder: C:\qxr\repo\conan-center-index\recipes\toon\all\test_package\build\msvc-193-x86_64-17-release\generators
toon/3.2 (test package): CMakeToolchain generated: conan_toolchain.cmake
toon/3.2 (test package): CMakeToolchain generated: C:\qxr\repo\conan-center-index\recipes\toon\all\test_package\build\msvc-193-x86_64-17-release\generators\CMakePresets.json
toon/3.2 (test package): CMakeToolchain generated: C:\qxr\repo\conan-center-index\recipes\toon\all\test_package\CMakeUserPresets.json
toon/3.2 (test package): Generating aggregated env files
toon/3.2 (test package): Generated aggregated env files: ['conanrun.bat', 'conanbuild.bat']

======== Testing the package: Building ========
toon/3.2 (test package): Calling build()
toon/3.2 (test package): Running CMake.configure()
toon/3.2 (test package): RUN: cmake -G "Ninja Multi-Config" -DCMAKE_TOOLCHAIN_FILE="generators/conan_toolchain.cmake" -DCMAKE_INSTALL_PREFIX="C:/qxr/repo/conan-center-index/recipes/toon/all/test_package" -DCMAKE_POLICY_DEFAULT_CMP0091="NEW" "C:/qxr/repo/conan-center-index/recipes/toon/all/test_package"     
conanvcvars.bat: Activating environment Visual Studio 17 - amd64 - winsdk_version=None - vcvars_ver=14.3
[vcvarsall.bat] Environment initialized for: 'x64'
-- Using Conan toolchain: C:/qxr/repo/conan-center-index/recipes/toon/all/test_package/build/msvc-193-x86_64-17-release/generators/conan_toolchain.cmake
-- Conan toolchain: C++ Standard 17 with extensions OFF
-- The CXX compiler identification is MSVC 19.39.33523.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/2022/Professional/VC/Tools/MSVC/14.39.33519/bin/Hostx64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found PkgConfig: C:/Users/edukaj/.conan2/p/b/pkgco8e02b502df764/p/bin/pkgconf.exe (found version "2.1.0")
-- Checking for module 'TooN'
--   Found TooN, version 3.2
-- Configuring done (2.9s)
-- Generating done (0.0s)
-- Build files have been written to: C:/qxr/repo/conan-center-index/recipes/toon/all/test_package/build/msvc-193-x86_64-17-release

toon/3.2 (test package): Running CMake.build()
toon/3.2 (test package): RUN: cmake --build "C:\qxr\repo\conan-center-index\recipes\toon\all\test_package\build\msvc-193-x86_64-17-release" --config Release -- -j8
conanvcvars.bat: Activating environment Visual Studio 17 - amd64 - winsdk_version=None - vcvars_ver=14.3
[vcvarsall.bat] Environment initialized for: 'x64'
[2/2] Linking CXX executable Release\test_package.exe


======== Testing the package: Executing test ========
toon/3.2 (test package): Running test()
toon/3.2 (test package): RUN: Release\test_package
determinant: 4.61
Lapack_Cholesky determinat: 4.61
```

---

- [X] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [X] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [ ] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/toon/all/conanfile.py [hunk 1]
```diff
@@ -57,7 +65,7 @@ def validate(self):
                 f"{self.ref} requires C++{self._min_cppstd}, which your compiler does not support."
             )
 
-        if not self.dependencies["openblas"].options.build_lapack:
+        if self.options.with_lapack and not self.dependencies["openblas"].options.build_lapack:
             raise ConanInvalidConfiguration("TooN requires LAPACK support in OpenBLAS")
```
Comment (RubenRBS) [inline recipes/toon/all/conanfile.py:69]:
```suggestion
            raise ConanInvalidConfiguration(f"{self.ref} requires LAPACK support in OpenBLAS with -o='openblas/*:build_lapack=True'")
```

**Review by RubenRBS — COMMENTED** [review]
Thanks a lot, this looks great - just a minor wording suggestion.

Also, could you please re-provide the compilation logs for `-o="toon/*:with_lapack=False"` but adding `--build="toon/*"` so that we can see the actual compilation taking place? Thanks!

**Comment by elvisdukaj** [comment]
Hi @RubenRBS , thank you for your feedback! Below the full log:

```
======== Exporting recipe to the cache ========
toon/3.2: Exporting package recipe: C:\qxr\repo\conan-center-index\recipes\toon\all\conanfile.py
toon/3.2: exports: File 'conandata.yml' found. Exporting it...
toon/3.2: Calling export_sources()
toon/3.2: Copied 1 '.yml' file: conandata.yml
toon/3.2: Copied 1 '.py' file: conanfile.py
toon/3.2: Copied 1 '.hh' file: config.hh
toon/3.2: Copied 1 '.patch' file: 001-bugfixes-backport.patch
toon/3.2: Exported to cache folder: C:\Users\edukaj\.conan2\p\toon8651ab841d614\e
toon/3.2: Exported: toon/3.2#2ec368e97ee6412fe866134e728f8a83 (2024-06-05 12:50:07 UTC)

======== Input profiles ========
Profile host:
[settings]
arch=x86_64
build_type=Release
compiler=msvc
compiler.cppstd=17
compiler.runtime=dynamic
compiler.runtime_type=Release
compiler.version=193
os=Windows
[options]
with_lapack=False
toon/*:with_lapack=False
[tool_requires]
*: ninja/1.11.1, cmake/3.28.1
[conf]
tools.cmake.cmaketoolchain:generator=Ninja Multi-Config
tools.env.virtualenv:powershell=False

Profile build:
[settings]
arch=x86_64
build_type=Release
compiler=msvc
compiler.cppstd=20
compiler.runtime=dynamic
compiler.runtime_type=Release
compiler.version=193
os=Windows
[conf]
tools.cmake.cmaketoolchain:generator=Ninja Multi-Config
tools.env.virtualenv:powershell=False


======== Computing dependency graph ========
Graph root
    cli
Requirements
    toon/3.2#2ec368e97ee6412fe866134e728f8a83 - Cache
Build requirements
    cmake/3.28.1#1a5bd0a27293689fd6fa44584e8e2322 - Cache
    ninja/1.11.1#cd30ca226f2040a979c359ebca4817d5 - Cache

======== Computing necessary packages ========
toon/3.2: Forced build from source
Requirements
    toon/3.2#2ec368e97ee6412fe866134e728f8a83:da39a3ee5e6b4b0d3255bfef95601890afd80709 - Build
Build requirements
    cmake/3.28.1#1a5bd0a27293689fd6fa44584e8e2322:522dcea5982a3f8a5b624c16477e47195da2f84f#29e6903b5006c2b2f81003c9ebf7551b - Cache
    ninja/1.11.1#cd30ca226f2040a979c359ebca4817d5:723257509aee8a72faf021920c2874abc738e029#e9b7296675fbd45c935f2062d16e31a6 - Cache

======== Installing packages ========
cmake/3.28.1: Already installed! (1 of 3)
cmake/3.28.1: Appending PATH environment variable: C:\Users\edukaj\.conan2\p\cmake689de4a54c3e9\p\bin
cmake/3.28.1: Appending PATH environment variable: C:\Users\edukaj\.conan2\p\cmake689de4a54c3e9\p\bin
ninja/1.11.1: Already installed! (2 of 3)
toon/3.2: Calling source() in C:\Users\edukaj\.conan2\p\toon8651ab841d614\s\src

-------- Installing package toon/3.2 (3 of 3) --------
toon/3.2: Building from source
toon/3.2: Package toon/3.2:da39a3ee5e6b4b0d3255bfef95601890afd80709
toon/3.2: Copying sources to build folder
toon/3.2: Building your package in C:\Users\edukaj\.conan2\p\b\toona748465b773aa\b
toon/3.2: Generating aggregated env files
toon/3.2: Generated aggregated env files: ['conanbuild.bat', 'conanrun.bat']
toon/3.2: Calling build()
toon/3.2: Apply patch (backport): apply unreleased minor fixes
toon/3.2: Package 'da39a3ee5e6b4b0d3255bfef95601890afd80709' built
toon/3.2: Build folder C:\Users\edukaj\.conan2\p\b\toona748465b773aa\b\build-release
toon/3.2: Generating the package
toon/3.2: Packaging in folder C:\Users\edukaj\.conan2\p\b\toona748465b773aa\p
toon/3.2: Calling package()
toon/3.2: package(): Packaged 33 '.h' files
toon/3.2: package(): Packaged 19 '.hh' files
toon/3.2: package(): Packaged 1 '.in' file: config.hh.in
toon/3.2: package(): Packaged 1 file: COPYING
toon/3.2: Created package revision ab90657a13138c4a6a8966dda4510e08
toon/3.2: Package 'da39a3ee5e6b4b0d3255bfef95601890afd80709' created
toon/3.2: Full package reference: toon/3.2#2ec368e97ee6412fe866134e728f8a83:da39a3ee5e6b4b0d3255bfef95601890afd80709#ab90657a13138c4a6a8966dda4510e08
toon/3.2: Package folder C:\Users\edukaj\.conan2\p\b\toona748465b773aa\p
WARN: deprecated: Usage of deprecated Conan 1.X features that will be removed in Conan 2.X:
WARN: deprecated:     'env_info' used in: cmake/3.28.1

======== Launching test_package ========

======== Computing dependency graph ========
Graph root
    toon/3.2 (test package): C:\qxr\repo\conan-center-index\recipes\toon\all\test_package\conanfile.py
Requirements
    toon/3.2#2ec368e97ee6412fe866134e728f8a83 - Cache
Build requirements
    cmake/3.28.1#1a5bd0a27293689fd6fa44584e8e2322 - Cache
    meson/1.2.2#02e0f176bd8fc261bccce543fdd17dce - Cache
    ninja/1.11.1#cd30ca226f2040a979c359ebca4817d5 - Cache
    pkgconf/2.1.0#89cc09d86584f09a5fc7bd831df72196 - Cache

======== Computing necessary packages ========
Requirements
    toon/3.2#2ec368e97ee6412fe866134e728f8a83:da39a3ee5e6b4b0d3255bfef95601890afd80709#ab90657a13138c4a6a8966dda4510e08 - Cache
Build requirements
    cmake/3.28.1#1a5bd0a27293689fd6fa44584e8e2322:522dcea5982a3f8a5b624c16477e47195da2f84f#29e6903b5006c2b2f81003c9ebf7551b - Cache
    ninja/1.11.1#cd30ca226f2040a979c359ebca4817d5:723257509aee8a72faf021920c2874abc738e029#e9b7296675fbd45c935f2062d16e31a6 - Cache
    pkgconf/2.1.0#89cc09d86584f09a5fc7bd831df72196:43771b8671ac44479c188dd72670e2eb2d7918a6#fa87fab9b26033fa9723c1a48ed27985 - Cache
Skipped binaries
    meson/1.2.2

======== Installing packages ========
cmake/3.28.1: Already installed! (1 of 4)
cmake/3.28.1: Appending PATH environment variable: C:\Users\edukaj\.conan2\p\cmake689de4a54c3e9\p\bin
ninja/1.11.1: Already installed! (2 of 4)
toon/3.2: Already installed! (3 of 4)
pkgconf/2.1.0: Already installed! (4 of 4)
pkgconf/2.1.0: WARN: The use of 'unix_path_legacy_compat' is deprecated in Conan 2.0 and does not perform path conversions. This is retained for compatibility with Conan 1.x and will be removed in a future version.
WARN: deprecated: Usage of deprecated Conan 1.X features that will be removed in Conan 2.X:
WARN: deprecated:     'env_info' used in: cmake/3.28.1, pkgconf/2.1.0

======== Testing the package ========
Removing previously existing 'test_package' build folder: C:\qxr\repo\conan-center-index\recipes\toon\all\test_package\build\msvc-193-x86_64-17-release
toon/3.2 (test package): Test package build: build\msvc-193-x86_64-17-release
toon/3.2 (test package): Test package build folder: C:\qxr\repo\conan-center-index\recipes\toon\all\test_package\build\msvc-193-x86_64-17-release
toon/3.2 (test package): Writing generators to C:\qxr\repo\conan-center-index\recipes\toon\all\test_package\build\msvc-193-x86_64-17-release\generators
toon/3.2 (test package): Generator 'PkgConfigDeps' calling 'generate()'
toon/3.2 (test package): Generator 'VirtualRunEnv' calling 'generate()'
toon/3.2 (test package): Calling generate()
toon/3.2 (test package): Generators folder: C:\qxr\repo\conan-center-index\recipes\toon\all\test_package\build\msvc-193-x86_64-17-release\generators
toon/3.2 (test package): CMakeToolchain generated: conan_toolchain.cmake
toon/3.2 (test package): CMakeToolchain generated: C:\qxr\repo\conan-center-index\recipes\toon\all\test_package\build\msvc-193-x86_64-17-release\generators\CMakePresets.json
toon/3.2 (test package): CMakeToolchain generated: C:\qxr\repo\conan-center-index\recipes\toon\all\test_package\CMakeUserPresets.json
toon/3.2 (test package): Generating aggregated env files
toon/3.2 (test package): Generated aggregated env files: ['conanrun.bat', 'conanbuild.bat']

======== Testing the package: Building ========
toon/3.2 (test package): Calling build()
toon/3.2 (test package): Running CMake.configure()
toon/3.2 (test package): RUN: cmake -G "Ninja Multi-Config" -DCMAKE_TOOLCHAIN_FILE="generators/conan_toolchain.cmake" -DCMAKE_INSTALL_PREFIX="C:/qxr/repo/conan-center-index/recipes/toon/all/test_package" -DCMAKE_POLICY_DEFAULT_CMP0091="NEW" "C:/qxr/repo/conan-center-index/recipes/toon/all/test_package"
conanvcvars.bat: Activating environment Visual Studio 17 - amd64 - winsdk_version=None - vcvars_ver=14.3
[vcvarsall.bat] Environment initialized for: 'x64'
-- Using Conan toolchain: C:/qxr/repo/conan-center-index/recipes/toon/all/test_package/build/msvc-193-x86_64-17-release/generators/conan_toolchain.cmake
-- Conan toolchain: C++ Standard 17 with extensions OFF
-- The CXX compiler identification is MSVC 19.39.33523.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/2022/Professional/VC/Tools/MSVC/14.39.33519/bin/Hostx64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found PkgConfig: C:/Users/edukaj/.conan2/p/b/pkgco8e02b502df764/p/bin/pkgconf.exe (found version "2.1.0") 
-- Checking for module 'TooN'
--   Found TooN, version 3.2
-- Configuring done (3.4s)
-- Generating done (0.0s)
-- Build files have been written to: C:/qxr/repo/conan-center-index/recipes/toon/all/test_package/build/msvc-193-x86_64-17-release

toon/3.2 (test package): Running CMake.build()
toon/3.2 (test package): RUN: cmake --build "C:\qxr\repo\conan-center-index\recipes\toon\all\test_package\build\msvc-193-x86_64-17-release" --config Release -- -j8
conanvcvars.bat: Activating environment Visual Studio 17 - amd64 - winsdk_version=None - vcvars_ver=14.3
[vcvarsall.bat] Environment initialized for: 'x64'
[2/2] Linking CXX executable Release\test_package.exe


======== Testing the package: Executing test ========
toon/3.2 (test package): Running test()
toon/3.2 (test package): RUN: Release\test_package
determinant: 4.61

```

**Review by RubenRBS — APPROVED** [review]
Thanks a lot!

**Review by jcar87 — APPROVED** [review]
Thanks!!

The way this recipe is handling `config.hh` is rather unfortunate - as it is hardcoding thigns that will have to be changed as new versions are added. Lapack was actually optional here from the beginning

**Pull request merged**